Add missing session id and tlsext_status accessors

rgacogne opened the pull request:
This commit adds two accessors for session id / context, used by haproxy's session cache:
* SSL_SESSION_set1_id()
* SSL_SESSION_get_id_context()

It also adds two accessors to retrieve the current value of the tlsext status callback and associated arg:
* SSL_CTX_get_tlsext_status_cb()
* SSL_CTX_get_tlsext_status_arg()

richsalz commented:
we need docs for these.  at least one-liners.  they can go into existing pod files.  and run util/find-doc-nits  script.

mattcaswell commented:
Looks like the requested docs have been added.

rgacogne commented:
Yes, I expected to be able to write even a bit more before adding a comment but couldn't find the time, sorry about that.

dwmw2 commented:
Can we set the protocol version, cipher and master key too? Should we be able to?

In OpenConnect for OpenSSL 1.1 I just switched¹ to generating the session information as ASN.1 and using `d2i_SSL_SESSION()`, but perhaps that isn't the best answer?

¹ http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/5abb133f

richsalz commented:
I think the d2i approach is the best thing to do for this release.

rgacogne commented:
Just for the record, haproxy already uses `d2i_SSL_SESSION` and `i2d_SSL_SESSION`, but at some point someone decided to set the session id to NULL before serializing to save a bit of memory, since we already keep it in our own struct. I guess I could break compatibility when compiled with OpenSSL 1.1+ if needed, but I'd like to avoid that if possible.

richsalz commented:
Can you provide more detail, Remi?

rgacogne commented:
Sure. Just to clarify, this is about the reason why I'd like to have something like `SSL_SESSION_set1_id()` to keep the current behavior of haproxy.
The current haproxy code sets `session_id_length` to 0 before calling `i2d_SSL_SESSION()` to serialize the SSL session in the session store. The comment that goes along this line states:

> Session id is already stored in to key and session id is known so we dont store it to keep size

The initial value of `session_id_length` and the session ID are restored when the session is retrieved from the cache, after `d2i_SSL_SESSION()` has been called. 
So without something like `SSL_SESSION_set1_id()`, I don't think I can keep the existing behavior when compiled against OpenSSL 1.1+. I hope this is clearer now, but please let me know otherwise!

dwmw2 commented:
https://github.com/haproxy/haproxy/blob/master/src/shctx.c#L388

They artificially clear the session_id in the SSL_SESSION, before calling i2d_SSL_SESSION() and storing the result. And later (#L503) they put it back again after calling d2i_SSL_SESSION().

They might be better off doing *all* that messing around in the serialised ASN.1 instead of in the SSL_SESSION itself....

mattcaswell commented on ssl/ssl_sess.c:
```diff
@@ -240,6 +240,13 @@ const unsigned char *SSL_SESSION_get_id(const SSL_SESSION *s,
         *len = s->session_id_length;
     return s->session_id;
 }
+const unsigned char *SSL_SESSION_get_id_context(const SSL_SESSION *s,
```
Probably should be get0?

mattcaswell commented on ssl/ssl_sess.c:
```diff
@@ -240,6 +240,13 @@ const unsigned char *SSL_SESSION_get_id(const SSL_SESSION *s,
+const unsigned char *SSL_SESSION_get_id_context(const SSL_SESSION *s,
+                                                unsigned int *len)
+{
+    if (len)
```
Make it `if (len != NULL)`

mattcaswell commented on doc/ssl/SSL_CTX_set_tlsext_status_cb.pod:
```diff
@@ -18,7 +20,11 @@ SSL_set_tlsext_status_ocsp_resp
+                                   int (**callback)(SSL *, void *));
+
  long SSL_CTX_set_tlsext_status_arg(SSL_CTX *ctx, void *arg);
+ long SSL_CTX_get_tlsext_status_arg(SSL_CTX *ctx, void **arg);
```
These need to be explained in the DESCRIPTION below.

mattcaswell commented on doc/ssl/ssl.pod:
```diff
@@ -441,6 +443,8 @@ sessions defined in the B<SSL_SESSION> structures.
 
 =item int B<SSL_SESSION_set_ex_data>(SSL_SESSION *s, int idx, char *arg);
 
+=item int B<SSL_SESSION_set1_id>(SSL_SESSION *s, const unsigned char *sid, unsigned int sid_len);
```
These need to be added to their own pod pages and properly described.

mattcaswell commented:
Ping @rgacogne. There are some outstanding comments on this MR to deal with. I'd like to get this in for the 1.1.0 release - but that is only 9 days away!

rgacogne commented:
* Rebased to fix conflicts
* Renamed SSL_SESSION_get_id_context -> SSL_SESSION_get0_id_context
* Made the NULL comparison explicit
* Added the missing documentation bits

mattcaswell commented:
+1

mattcaswell commented:
Ping any team member for second review.

levitte commented:
+1

levitte commented:
(to be noted the decrease of coverage.  for another PR)

mattcaswell commented:
Merged! Thanks.
